Source/WebKit/webpushd/WebPushDaemon.mm

```diff
@@ -736,12 +736,53 @@ static bool platformShouldPlaySound(const WebCore::NotificationData& data)
             default: return WebCore::PushPermissionState::Prompt;
             }
         }(settings.authorizationStatus);
-        replySender(permissionState);
+        RELEASE_LOG(Push, "getPushPermissionState for %{sensitive}s with result: %u", originString.utf8().data(), static_cast<unsigned>(permissionState));
+
+        WorkQueue::main().dispatch([replySender = WTFMove(replySender), permissionState] mutable {
+            replySender(permissionState);
+        });
     });
 
     [center getNotificationSettingsWithCompletionHandler:blockPtr.get()];
 }
 
+void WebPushDaemon::requestPushPermission(PushClientConnection& connection, const WebCore::SecurityOriginData& origin, CompletionHandler<void(bool)>&& replySender)
+{
+    auto identifier = connection.subscriptionSetIdentifier();
+    if (identifier.pushPartition.isEmpty()) {
+        WEBPUSHDAEMON_RELEASE_LOG_ERROR(Push, "Denied push permission since no pushPartition specified");
+        return replySender(false);
+    }
+
+#if PLATFORM(IOS)
+    const auto& webClipIdentifier = identifier.pushPartition;
+    RetainPtr webClip = [UIWebClip webClipWithIdentifier:(NSString *)webClipIdentifier];
+    auto webClipOrigin = WebCore::SecurityOriginData::fromURL(URL { [webClip pageURL] });
+
+    if (origin.isNull() || origin.isOpaque() || origin != webClipOrigin) {
+        WEBPUSHDAEMON_RELEASE_LOG(Push, "Denied push permission because web clip origin %{sensitive}s does not match expected origin %{sensitive}s", webClipOrigin.toString().utf8().data(), origin.toString().utf8().data());
+        return replySender(false);
+    }
+#endif
+
+    NSString *notificationCenterBundleIdentifier = platformNotificationCenterBundleIdentifier(connection);
+    RetainPtr center = adoptNS([[UNUserNotificationCenter alloc] initWithBundleIdentifier:notificationCenterBundleIdentifier]);
```

Comment:
Can we test this right now with the mock Notification Center? (Which would also mean allocating the current Notification Center class here instead of UNUserNotificationCenter)